### PR TITLE
Threadedplanner

### PR DESCRIPTION
**Pull request by atdotde**
### Describe the pull request:

- [ ] Bug fix
- [x] Functional change
- [x] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This is the big refactor to remove global state variables of the planner/deco parts. Everything
is now in struct deco_state and pointers to it get passed around. This is necessary to have several
instances of the planner run in parallel (as required for UI responsiveness).

Unfortunately, a variable rename got squashed into the same commit.

### Changes made:

Added struct deco_state *ds arguments to many functions.
Renamed instances from deco_state to ds.

### Mentions: 
@RickMWalsh

### CODE REVIEW

**Comment by janmulder** [comment]
@atdotde: not sure how many relevant testing can be done without rebasing ... tried a quick rebase myself, but better that you and/or @RickMWalsh rebase things.

**Comment by atdotde** [comment]
This is now relative to current master.

**Comment by janmulder** [comment]
Does not compile. Obviously, I could do a PR on your PR ... but might be easier that ypu fix the trivial missing renames to "ds".

That said. I fixed it myself, and found something nasty.

Access menu-log-add dive. Application hangs on` planLock.lock() `in the UI thread

**Comment by janmulder** [comment]
@atdotde: tried to do a PR to your PR to fix the compile. Just merge it to your branch and it is supposed to show up here. (GitHub magic :-) )

**Comment by RickMWalsh** [comment]
@atdotde I've tested the planner in your branch (with @janmulder 's ds patch applied).  The UI appears considerably more responsive when choosing depth and duration by dragging points around in the profile.  But I found a few issues.
- Variations don't display, just the word "VARIATIONS"
- For some planned dives (especially with long bottom times), it appears the deco calculations are skipped on re-planning when preferences are changed, or bottom time is changed, but not when depth is changed in the table.  E.g. see ascent directly to the surface. 
image
- Displayed VPM-B ceiling is under-estimated (but VPM-B plans are unchanged from before), both in planner and for real dives (in limited testing)

Cheers,

Rick

**Comment by janmulder** [comment]
> That said. I fixed it myself, and found something nasty.
> 
> Access menu-log-add dive. Application hangs onplanLock.lock()in the UI thread

And pushed a commit for this as well.

**Comment by atdotde** [comment]
Stupid me for pushing something that does not build. I was in a rush and concluded that if there are no more merge conflicts things should be fine. Sorry @janmulder, I could not find your PR, but amended my branch in the obvious way (I guess you did the same things).

@RickMWalsh So far, there should be no speed improvement in the UI yet, since there is no parallelism (but only preparation for it). Only that, as you notice, the variations are temporary commented out. Those should be done in the background. I just wanted to get this out so people can touch the planner source again. I have no clue how the VPM-B ceiling could be affected but will investigate this. Do you have an example file where it is obvious?

**Comment by janmulder** [comment]
@atdotde: look at your github account :-) you should see an incoming PR. (https://github.com/atdotde/subsurface/pull/2)

**Review by janmulder — COMMENTED** [review]
There is second return in plan() that probably needs this.

**Comment by atdotde** [comment]
OK, I am learning...

**Comment by janmulder** [comment]
Ok, things are now in a more or less sane state. What is the plan now? Leave this open as WIP or merge it, and open issues or PRs for remaining work?

**Comment by atdotde** [comment]
No, the idea is to merge it as soon as we are somewhat confident that the breakage is not too big. Otherwise it would either prevent other independent work on too many files or would be a nightmare to merge later. At least that's my suggestion.

And it is of course my plan to continue here along these lines and actually get to a multi-threaded planner (including much finer lock granularity). But my experience with these kinds of things is very limited so progress might be very slow.

**Comment by sfuchs79** [comment]
I also started some testing and can mainly confirm @RickMWalsh findings.
I found an interesting example (xml attached).
If you change in the example planned dive the bottom time between 59 and 60 minutes the resulting ceiling makes a huge jump which is for sure not correct:

grafik

grafik

[Planner_Ceiling_1.zip](https://github.com/Subsurface-divelog/subsurface/files/1496729/Planner_Ceiling_1.zip)

**Comment by dirkhh** [comment]
> the idea is to merge it as soon as we are somewhat confident that the breakage is not too big. 

I usually request a _slightly_ higher standard than that. Remember, the idea is that we continue to be able to do a roughly monthly release - so I'd do a new release in a week or two, right? So it would be good if we were free of, let's call it "major regressions" :-)
I understand the idea here, I understand the desire to get it into master so it doesn't become a maintenance nightmare. But I'd also like to see it be usable in a 4.7.5
@janmulder @RickMWalsh - can you comment on how close we are to that and what would need to be fixed?

**Comment by janmulder** [comment]
@dirkhh: not really an opinion on this. That is, with the mutex blockers fixed (and compiling again), and some random clicking around in the planner (that I basically never use) I do not see major strange things that would prevent merge. But not looking at specific VPM-B stuff (like Rick and Stefan), so my perspective on this is limited.

File: core/profile.c
Comment (sfuchs79) [inline core/profile.c]: I guess calling `restore_deco_state(cache_data_initial, ds, true);` here for a second time is not correct.

File: core/profile.c
Comment (sfuchs79) [inline core/profile.c]: Does one really need the additional `struct deco_state *ds` pointer here? One can pass &plot_deco_state later on.

File: core/profile.c
Comment (sfuchs79) [inline core/profile.c]: This doesn't look that nice from coding style point of view.

File: core/profile.c
Comment (sfuchs79) [inline core/profile.c:1060]: Just out of curiosity: Does this give a compiler warning w/o the parentheses?

**Review by sfuchs79 — CHANGES_REQUESTED** [review]
Fixing the one thing with the duplicated `restore_deco_state` call changes/improves the ceiling results slightly but unfortunately doesn't fix things completely. Must be another small mistake somewhere...

**Comment by sfuchs79** [comment]
Ah, no, mistake done when testing. Not sure any more if my statement about the `restore_deco_state(cache_data_initial, ds, true);` thing is even correct...

File: core/profile.c
Comment (atdotde) [reply core/profile.c:1060]: yes. clang does on Mac.

**Comment by atdotde** [comment]
Implemented @sfuchs79 's suggestions. Will look at the problem that @RickMWalsh mentioned. Hope this is ready to merge in time. This branch should really be a functional no-op.

**Comment by RickMWalsh** [comment]
@dirkhh while I tested the planner last night, I haven't had a chance to look at the code at all so I'm not in a good position to comment on how much effort is required to fix the regressions.  Certainly they need fixing before the next release, or the next release needs to be done before merging.  My experience in software maintenance is zero, but have there been substantial changes in Master affecting the desktop program since the last point release and/or substantial changes imminent?  If so, is a point release off current (or soon to be current) master possible, then merge this pull request warts-and-all?  Alternatively, if this looks to be the most significant pending change, maybe it's worth merging now and holding off the release until the bugs are ironed out.  Looking through the commit log, it looks like recent commits are primarily typos, mobile, and build/travis/github fixes.  All are definitely worthy, but mostly not noticeable to a desktop user of the next point release.
@atdotde can you reproduce and try to work out why changing the waypoint time in the planner, or planner settings sometimes fails to trigger the deco calculations when calculating the ascent (but it seems only for reasonably long bottom/deco times)?  I noticed it using Buhlmann and @sfuchs79 noticed the same thing with VPM-B, so it doesn't appear to be algorithm specific.
For the VPM-B under-estimated ceiling, this looks similar to the recent issues where deco_time is either miscalculated, not carried over from the actual plan, or calculated "correctly" but not handled properly when calculating the ceiling.  If this isn't fixed before, I'll try to have a look at the code over the weekend.
Has anyone tested recreational mode?

**Comment by atdotde** [comment]
I just pushed a commit even though it does not build to ask @tcanabrava  for help. This should eventually address the vpmb profile shown in planned dives.

**Comment by dirkhh** [comment]
Looking at Travis, Linux, Mac, and Windows seem to build, Android stumbles over an undefined symbol. I can easily fix that if you want. Why were you specifically calling out @tcanabrava - he's usually the Qt guy, but I didn't see any failures there. Did I miss something?

**Comment by dirkhh** [comment]
@atdotde Here's the patch that fixes the build of Subsurface-mobile for Android for me
```
diff --git a/core/profile.c b/core/profile.c
index d521c1ef0e7b..0ead58e85b28 100644
--- a/core/profile.c
+++ b/core/profile.c
@@ -1306,10 +1306,11 @@ void create_plot_info_new(struct dive *dive, struct divecomputer *dc, struct plo
 #ifndef SUBSURFACE_MOBILE
        struct deco_state plot_deco_state;
        init_decompression(&plot_deco_state, dive);
-#endif
+
        /* Create the new plot data */
        if(in_planner())
                plot_deco_state = final_plan_ds();
+#endif
        free((void *)last_pi_entry_new);
 
        get_dive_gas(dive, &o2, &he, &o2max);
```
Of course, looking at this, the Linux build of Subsurface-mobile should also have failed. Have to look at the Travis output again :-)

**Comment by dirkhh** [comment]
facepalm - Travis doesn't realize that the builds actually fail. That's not good.
So two things I need to look into
- what's up with the Travis build and why doesn't Travis see that it's failing
- why does a local build fail with ` undefined reference to 'DivePlannerPointsModel::instance()'`
I assume the latter is the linker problem you mentioned, @atdotde ?

It's Thanksgiving in the US today - I just snuck away for 20 minutes to look at pull requests... I doubt I'll find time to deal with this today but I'll definitely look into it tomorrow (assuming no one else does so, first)

**Comment by atdotde** [comment]
After a long discussion with @tcanabrava on IRC, I approached this in a different way (by passing down parameters). This avoids the linker problem (but we should make sure Travis detects those in the future). 

In my quick testing, this fixes the VPM-B ceiling in the planner. Could somebody please double check?

Could somebody please check? 

I will look into the "planner stops updating" problem soon. But I think I am getting closer to get this to a mergeable state. Once ready, I will do a final rebase -i to sort everything into a sane set of commits.

File: core/planner.c
Comment (sfuchs79) [inline core/planner.c]: Coding style

File: core/planner.c
Comment (sfuchs79) [inline core/planner.c]: This also may need some cleanup in terms of line breaks,...

File: core/profile.c
Comment (sfuchs79) [inline core/profile.c:920]: Again

File: core/profile.c
Comment (sfuchs79) [inline core/profile.c:938]: Again

File: core/profile.c
Comment (sfuchs79) [inline core/profile.c:955]: Again

File: qt-models/diveplannermodel.cpp
Comment (sfuchs79) [inline qt-models/diveplannermodel.cpp:971]: Why is this change needed?

**Review by sfuchs79 — COMMENTED** [review]
Did another review of the changed code and found a few more coding style things plus one change where I have a question. Beside this LGTM - no other suspicious things found.

**Comment by sfuchs79** [comment]
Even more important I also did some testing with good results:
- Tested my small set of planned dives in VPM-B mode and compared the results with results from current master -> all same and ok.
- Logged dive also look ok (also VPM-B +2)
- Can't say anything  about the "deco calculations skipped" item from @RickMWalsh because I was never able to reproduce this one

**Comment by dirkhh** [comment]
Thanks for the great review @sfuchs79  - I think we are getting closer to being able to pull this.
I'm working on the Travis issues with builds not failing when they should... that will increase the confidence in these changes as well... (because right now the Linux build fails without showing it as failed, but that's a completely unrelated bug that has been in master for over a week :-( )

**Comment by dirkhh** [comment]
I restarted builds now that they appear to a) complete again on master and b) fail when they don't complete

**Comment by dirkhh** [comment]
Tests now look good. @atdotde - is this ready to be merged? @sfuchs79 ?

**Comment by RickMWalsh** [comment]
Without looking at the code in any detail, I've tested the planner (Buhlmann, VPM-B and rec mode) and gone through a number of real and planned dives in my dive list.  I can find no difference in the calculated plans or displayed profile ceilings compared to 4.7.4.  I can no longer reproduce the previous bug of deco calculations being "missed" when planner settings are changed. So looks good to me too.

The only regression/change I can see is that variations have been disabled.  Is it possible to re-enable them now?

**Comment by dirkhh** [comment]
Thanks for testing @RickMWalsh - that sounds very promising.
@atdotde - up to you if you want to merge it.

File: qt-models/diveplannermodel.cpp
Comment (atdotde) [reply qt-models/diveplannermodel.cpp:971]: Otherwise, there is potential crash if if last_segment->next is NULL upon trying to access NULL->next.

File: core/planner.c
Comment (atdotde) [reply core/planner.c]: I tried to improve these. Some where caused by ds originally being called deco_state. But in order to somewhat stay within reasonable number of columns there have to be breaks and in the following line we need enough indentation.

**Comment by atdotde** [comment]
This final force-push is a rebase -i of what we had before plus a bit of white space clean up in long lines. 

Thank you everybody for your help!

**Comment by sfuchs79** [comment]
@atdotde 
Robert, one more question about your plans for this:
- Will it finally be possible to run ANY calls of plan(); - also the ones for the "primary" dive plan calculation - in a separate thread? I mean not only the ones for the plan variations?
I'm asking this because I already confirmed with some tests today that on very slow HW (e.g. my Atom based netbook) the old "double increment/decrement" issue #602 does not disappear after these changes yet. I believe the computation time for a single plan is still too long to avoid the UI issue.

**Comment by atdotde** [comment]
So far, all I have done are preparations. Nothing actually runs in a separate thread. The main plan is slightly more complicated than the variations since we need to figure out how the signalling to the profile once the plan is complete is supposed to work.

**Pull request merged**
